File: Data Structures & Algorithms/depth-of-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {
        if (!root) {
            return 0;
        }

        const leftHeight = this.maxDepth(root.left);
        const rightHeight = this.maxDepth(root.right);

        return 1 + (Math.max(leftHeight, rightHeight));
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # initiate res, and define dfs, dont forget nonlocal res
        # if root is none return 0, recursively do dfs to left and right of root
        # then update res to be max between res and the sum of left and right
        # return 1 + max of left or right. then run dfs on root, and return res

        res = 0

        def dfs(root):
            nonlocal res

            if root is None:
                return 0

            right = dfs(root.right)
            left = dfs(root.left)

            res = max(res, right + left)
            return 1 + max(left, right)
        
        dfs(root)

        return res